### PR TITLE
GetFiles and GetDirectories return names instead of paths

### PR DESCRIPTION
**Pull request by jorolf**
Before this change they would've returned a path to the file/directory making it harder to e.g. reuse it for ``Delete`` or ``GetStream``

---

### CODE REVIEW

**Comment by peppy** [comment]
This is going to require an update on osu!'s end, as it breaks compatibility. 

https://github.com/ppy/osu/blob/0704f962c51578bd129b31ac93b329672a11e0f8/osu.Game/Beatmaps/BeatmapManager.cs#L332.

Also i'm not sure about this change for subdirectories. Correct me if I'm wrong but it's always returning *just* the filename, not the relative path. I think we may want to return relative path if anything (especially if you're using the rationale that the output can be piped into other `Storage` functions directly)?

File: osu.Framework/Platform/DesktopStorage.cs [hunk 1]
```diff
@@ -34,9 +36,9 @@ public override void DeleteDirectory(string path)
 
         public override void Delete(string path) => FileSafety.FileDelete(GetUsablePathFor(path));
 
-        public override string[] GetDirectories(string path) => Directory.GetDirectories(GetUsablePathFor(path));
+        public override IEnumerable<string> GetDirectories(string path) => Directory.GetDirectories(GetUsablePathFor(path)).Select(str => str.Substring(GetUsablePathFor("").Length + 1)); // the last path separator character has to be removed, too
```
Comment (peppy) [inline osu.Framework/Platform/DesktopStorage.cs]:
```csharp
public override IEnumerable<string> GetDirectories(string path) =>Directory.GetDirectories(GetUsablePathFor(path)).Select(getRelativePath);

private string getRelativePath(string path)
{
    string basePath = GetUsablePathFor("");
    if (!path.Contains(basePath)) throw new ArgumentException("Paths must have a common prefix");
    return path.Replace(basePath, "").TrimStart(Path.DirectorySeparatorChar);
}
```

might be a bit more safe and readable. haven't tested, so please do so.

Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: basePath is an absolute path, so should not be dangerous.

Comment (smoogipoo) [reply osu.Framework/Platform/DesktopStorage.cs]: Except in the case that you're on a unix system and the base path is maybe, `/Games`

Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: I guess so, in a very rare care where the relative path requested matches the absolute path. Substring can also be weird since it may incorrectly match if the sources of the absolute paths differ (ie. on windows you can have `\\` prefixes that still resolve the same filename).

We're doing this in `WindowsStorage`, so special care should be taken here. I believe the current implementation in this PR may break.

Comment (smoogipoo) [reply osu.Framework/Platform/DesktopStorage.cs]: Replace can't be used in that scenario either ;) Good catch though, definitely a serious issue.

Comment (smoogipoo) [reply osu.Framework/Platform/DesktopStorage.cs]: Upon discussion, we should be able to to wrap both paths inside `getRelativePaths()` with https://msdn.microsoft.com/en-us/library/system.io.path.getfullpath(v=vs.110).aspx, prior to performing the substring.

Comment (jorolf) [reply osu.Framework/Platform/DesktopStorage.cs]: Should I leave the ``path.Contains(basePath)``?

File: osu.Framework/Platform/Storage.cs [hunk 2]
```diff
@@ -34,18 +35,25 @@ protected Storage(string baseName)
...
             if (createIfNotExisting) Directory.CreateDirectory(Path.GetDirectoryName(resolvedPath));
             return resolvedPath;
         }
 
+        /// <summary>
+        /// Get a usable filesystem path for the provided incomplete path. The same as calling <see cref="GetUsablePathFor(string,bool)"/>.
+        /// </summary>
+        /// <param name="path">An incomplete path, usually provided as user input.</param>
+        /// <returns>A usable filesystem path.</returns>
+        public string GetUsablePathFor(string path) => GetUsablePathFor(path, false);
```
[the remark this answers is not in the document]
Comment (jorolf) [reply osu.Framework/Platform/Storage.cs]: The original is still protected while the addition is public

**Comment by peppy** [comment]
How have you been testing this? Can you add some tests for this?

at very least, the edge cases should be tested:

```
../    (should probably not work and i think it might currently)
.
./
subfolder/test/
subfolder/test
subfolder\test
```

**Comment by peppy** [comment]
Added testts and rearranged logic somewhat. Requesting a quick look @smoogipoo.

File: osu.Framework/Platform/DesktopStorage.cs [hunk 3]
```diff
@@ -19,30 +21,53 @@ public DesktopStorage(string baseName, GameHost host)
...
-        public override string[] GetFiles(string path) => Directory.GetFiles(GetUsablePathFor(path));
+        public override IEnumerable<string> GetFiles(string path) => getRelativePaths(Directory.GetFiles(GetFullPath(path)));
 
-        public override void OpenInNativeExplorer() => host.OpenFileExternally(GetUsablePathFor(string.Empty));
+        private IEnumerable<string> getRelativePaths(IEnumerable<string> paths)
+        {
+            string basePath = Path.GetFullPath(GetFullPath(string.Empty));
+            return paths.Select(Path.GetFullPath).Select(path =>
+            {
+                if (!path.StartsWith(basePath)) throw new ArgumentException($"\"{path}\" does not start with \"{basePath}\" and is probably malformed");
```
Comment (jorolf) [inline osu.Framework/Platform/DesktopStorage.cs]: Isn't this check useless now because of line 60?

Comment (smoogipoo) [reply osu.Framework/Platform/DesktopStorage.cs]: Eh, I'd still rather have this here.

Comment (peppy) [reply osu.Framework/Platform/DesktopStorage.cs]: this is still required, as this function bypasses the check mentioned.

**Comment by smoogipoo** [comment]
Tested with osu! (requires a few changes), gtg.
